src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs

```diff
@@ -17,32 +19,24 @@
 
 namespace Microsoft.CodeAnalysis.FindSymbols
 {
-    // TODO : this can be all moved down to compiler side.
     internal static class FindReferenceCache
     {
-        private static readonly ReaderWriterLockSlim s_gate = new();
-        private static readonly Dictionary<SemanticModel, Entry> s_cache = new();
+        private static readonly ConditionalWeakTable<SemanticModel, Entry> s_cache = new();
+
+        private static Entry GetEntry(SemanticModel model)
+            => s_cache.GetValue(model, static model => new(model));
 
         public static SymbolInfo GetSymbolInfo(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
         {
-            var nodeCache = GetNodeCache(model);
-            if (nodeCache == null)
-            {
-                return model.GetSymbolInfo(node, cancellationToken);
-            }
+            var nodeCache = GetEntry(model).SymbolInfoCache;
```

Comment:
How much overhead does GetEntry end up costing? It seems like we could just do the lookup for the entry once and then pass it around to ProcessDocument paths therefore avoiding all the extra lookups? The ConditionalWeakTable can (potentially) stay but if we decided we didn't want it then it'd still mean you can have a cache for repeated lookups within the same FAR session but not across sessions.

Reply:
i was on hte fence about this, but i do like the pattern you're suggesting, so i'll move to that :)

Reply:
> but not across sessions.

Note: i'd really like to have cross-session caches.   I do repeated FARs a *lot*, and a bunch fo work was made here and in DependentTypeFinder/ProejctFinder to speed this up so that multiple-fars in a row can often be near free.  this is *really* helpful when trying to hunt things down in code.